File: backend/app/api/routes/workflows.py
"""Workflow API routes."""

from fastapi import APIRouter, Depends

from app.api.deps import get_workflow_service
from app.schemas.common import ApiResponse
from app.schemas.workflow import CreateWorkflowRequest, WorkflowArtifacts, WorkflowResponse
from app.services.workflow_service import WorkflowService

router = APIRouter()


@router.post("", response_model=ApiResponse[WorkflowResponse])
async def create_workflow(
    request: CreateWorkflowRequest,
    service: WorkflowService = Depends(get_workflow_service),
) -> ApiResponse[WorkflowResponse]:
    """Create a compliance workflow for an uploaded document."""
    workflow = service.create_workflow(document_id=request.document_id)
    return ApiResponse(success=True, data=workflow, error=None)


@router.get("/{workflow_id}", response_model=ApiResponse[WorkflowResponse])
async def get_workflow(
    workflow_id: str,
    service: WorkflowService = Depends(get_workflow_service),
) -> ApiResponse[WorkflowResponse]:
    """Return workflow status and stage outputs."""
    workflow = service.get_workflow(workflow_id)
    return ApiResponse(success=True, data=workflow, error=None)


@router.post("/{workflow_id}/run", response_model=ApiResponse[WorkflowResponse])
async def run_workflow(
    workflow_id: str,
    service: WorkflowService = Depends(get_workflow_service),
) -> ApiResponse[WorkflowResponse]:
    """Run the scaffolded compliance workflow."""
    workflow = service.run_workflow(workflow_id)
    return ApiResponse(success=True, data=workflow, error=None)


@router.get("/{workflow_id}/status", response_model=ApiResponse[WorkflowResponse])
async def get_workflow_status(
    workflow_id: str,
    service: WorkflowService = Depends(get_workflow_service),
) -> ApiResponse[WorkflowResponse]:
    """Return current workflow status for polling."""
    workflow = service.get_workflow(workflow_id)
    return ApiResponse(success=True, data=workflow, error=None)


@router.get("/{workflow_id}/artifacts", response_model=ApiResponse[WorkflowArtifacts])
async def get_workflow_artifacts(
    workflow_id: str,
    service: WorkflowService = Depends(get_workflow_service),
) -> ApiResponse[WorkflowArtifacts]:
    """Return planner, executor, and reviewer artifacts."""
    artifacts = service.get_artifacts(workflow_id)
    return ApiResponse(success=True, data=artifacts, error=None)

File: frontend/src/services/workflows.ts
import { apiGet, apiPost } from "./api";

export type WorkflowStageStatus = "pending" | "running" | "completed" | "failed" | "warning";
export type WorkflowStageName = "extraction" | "planner" | "executor" | "reviewer" | "revision" | "report";

export type WorkflowStageResult = {
  stage: WorkflowStageName;
  status: WorkflowStageStatus;
  execution_time_ms: number;
  provider_used: string;
  output_summary: string;
  issues: string[];
  progress: number;
};

export type WorkflowResponse = {
  workflow_id: string;
  document_id: string;
  status: WorkflowStageStatus;
  current_stage: WorkflowStageName;
  revision_count: number;
  stages: WorkflowStageResult[];
};

export type WorkflowArtifacts = {
  planner: Record<string, unknown>;
  executor: Record<string, unknown>;
  reviewer: Record<string, unknown>;
};

export function createWorkflow(documentId: string) {
  return apiPost<WorkflowResponse>("/api/workflows", { document_id: documentId });
}

export function getWorkflow(workflowId: string) {
  return apiGet<WorkflowResponse>(`/api/workflows/${workflowId}`);
}

export function runWorkflow(workflowId: string) {
  return apiPost<WorkflowResponse>(`/api/workflows/${workflowId}/run`);
}

export function getWorkflowStatus(workflowId: string) {
  return apiGet<WorkflowResponse>(`/api/workflows/${workflowId}/status`);
}

export function getWorkflowArtifacts(workflowId: string) {
  return apiGet<WorkflowArtifacts>(`/api/workflows/${workflowId}/artifacts`);
}

